fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/hud/HudElementRegistryImpl.java

```diff
@@ -0,0 +1,215 @@
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+	}
+
+	public static void replaceElement(Identifier identifier, Function<IdentifiedElement, IdentifiedElement> replacer) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.set(replacer.apply((IdentifiedElement) l));
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+	}
+
+	@VisibleForTesting
+	static void validateUnique(IdentifiedElement layer) {
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, layer.id())) {
+				throw new IllegalArgumentException("Layer with identifier " + layer.id() + " already exists");
+			}
+
+			return false;
+		});
+	}
+
+	/**
+	 * @return true if an element with the given identifier was found
+	 */
+	@VisibleForTesting
+	static boolean findLayer(Identifier identifier, ElementVisitor visitor) {
+		MutableBoolean found = new MutableBoolean(false);
+
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, identifier)) {
+				found.setTrue();
+				return visitor.visit(l, iterator);
+			}
+
+			return false;
+		});
+
+		return found.booleanValue();
+	}
+
+	@VisibleForTesting
+	static boolean visitLayers(ElementVisitor visitor) {
+		boolean modified = false;
+
+		for (VanillaElement vanillaElement : vanillaElements.sequencedValues()) {
+			modified |= visitLayers(vanillaElement.elements(), visitor);
+		}
+
+		return modified;
+	}
+
+	private static boolean visitLayers(List<HudElement> layers, ElementVisitor visitor) {
+		MutableBoolean modified = new MutableBoolean(false);
+		ListIterator<HudElement> iterator = layers.listIterator();
+
+		while (iterator.hasNext()) {
+			HudElement element = iterator.next();
+
+			if (visitor.visit(element, iterator)) {
+				modified.setTrue();
+			}
+		}
+
+		return modified.booleanValue();
+	}
+
+	private static boolean matchesIdentifier(HudElement element, Identifier identifier) {
+		return element instanceof IdentifiedElement ie && ie.id().equals(identifier);
+	}
+
+	@VisibleForTesting
+	interface ElementVisitor {
+		/**
+		 * @return true if the list has been modified, false if not modified
+		 */
+		boolean visit(HudElement element, ListIterator<HudElement> iterator);
+	}
+
+	/**
+	 * An element that wraps a vanilla element using a list, allowing for users to attach elements before or after it, replace it, or remove it.
+	 */
+	@VisibleForTesting
+	public record VanillaElement(Identifier id, List<HudElement> elements) {
+		public VanillaElement(Identifier id) {
+			this(id, new ArrayList<>());
+			elements().add(IdentifiedElement.of(id, (context, tickCounter) -> { }));
+		}
+
+		public void render(InGameHud instance, DrawContext context, RenderTickCounter tickCounter, Operation<Void> renderVanilla) {
```

Comment:
Using a Mixin Extras `Operation` feels a bit wrong outside of mixins.

Reply:
Changed this to pass a Runnable, mostly becuase some of hotbar renderes take diffrent args, and partly becuase it did feel odd.